crates/typst-library/src/meta/counter.rs

```diff
@@ -388,6 +388,29 @@ impl Counter {
         UpdateElem::new(self.0, update).pack()
     }
 
+    /// Calculate the difference between the real page number and the logical one.
+    ///
+    /// Difference between last counter change.
+    pub fn page_delta(&self, vt: &mut Vt, num: NonZeroUsize) -> SourceResult<usize> {
+        if !self.is_page() {
+            bail!(Span::detached(), "Only works for the page counter!");
+        }
+
+        let sequence = self.sequence(vt)?;
+        let offset = vt.introspector.query_up_to(num).len();
+
+        // Get the counter state (and therefore page) closest to the current physical page.
+        let (cs, page) = sequence
+            .iter()
+            .filter(|(_, p)| *p <= num)
+            .min_by(|(_, p1), (_, p2)| {
+                offset.abs_diff(p1.get()).cmp(&offset.abs_diff(p2.get()))
+            })
+            .unwrap();
```

Comment:
The idea is that the sequence length directly gives you the counter state closest to the current physical page. But that only works if the result of query_up_to actually matches the page counter selector (which also helps with incrementality).